### PR TITLE
Always run Assessment with ConfigurePatching - CRP/ARG issue mitigation

### PR DESCRIPTION
**Pull request by kjohn-msft**
**Reason for this change:**

Configurepatching results are wrapped inside assessment results at Azure Resource Graph. 

Which means that some assessment result needs to be written for the configurepatching results to be written. CRP can fake it – but none of the outcomes were deemed acceptable:
-	Writing success with defaults (0 security, 0 critical, etc) – this is misleading as it pretends the machine is secure. (Not good at scale either)
-	Writing warning with defaults (0 security, etc) – in current consumption experiences, this is also misleading. And the warning would be overwriting older but more accurate assessment results.
-	Writing failed with defaults – this will be considered a bug as assessment never failed. It was never attempted (or is still running). (At scale, some portion of a fleet will always have ‘failed’ assessment even though nothing failed)

If the Windows/Linux extensions attempt to write assessment (as ‘Platform’) with every configure patching result, this will always allow CRP to write accurate information to ARG (even if it fails at the extension).

**This PR mitigates the identified root cause in the Linux Patch Extension.**

Note: includes a change from another open PR that may merge before this one.

### CODE REVIEW

File: src/core/src/core_logic/ConfigurePatchingProcessor.py [hunk 1]
```diff
@@ -47,16 +48,28 @@ def start_configure_patching(self):
...
         return self.configure_patching_successful
 
+    def set_configure_patching_final_overall_status(self):
+        """ Writes the final overall status after any pre-requisite operation is also in a terminal state - currently this is only assessment """
+        overall_status = Constants.STATUS_SUCCESS if self.configure_patching_successful else Constants.STATUS_ERROR
+        if self.configure_patching_exception_error is None:
+            self.__report_consolidated_configure_patch_status(status=overall_status)
+        else:
+            self.__report_consolidated_configure_patch_status(status=overall_status, error=self.configure_patching_exception_error)
+
```
Comment (feng-j678) [inline src/core/src/core_logic/ConfigurePatchingProcessor.py:72]: doesn't this function's assertion of the status cover the above logic?

def test_patch_mode_set_failure_for_configure_patching(self):  
        self.assertTrue(substatus_file_data[0]["status"].lower() == Constants.STATUS_ERROR.lower())
        self.assertEqual(len(substatus_file_data), 2)
        self.assertTrue(substatus_file_data[0]["name"] == Constants.PATCH_ASSESSMENT_SUMMARY)   # assessment is 
        now part of the CP flow
        self.assertTrue(substatus_file_data[0]["status"].lower() == Constants.STATUS_SUCCESS.lower())
        self.assertTrue(substatus_file_data[1]["name"] == Constants.CONFIGURE_PATCHING_SUMMARY)
        self.assertTrue(substatus_file_data[1]["status"].lower() == Constants.STATUS_ERROR.lower())

File: src/core/src/CoreMain.py [hunk 2]
```diff
@@ -74,10 +74,10 @@ def __init__(self, argv):
...
 
-            # Assessment happens for an Auto Assessment request or for all Non Auto Assessment operations, except for ConfigurePatching iff AssessmentMode is set to AutomaticByPlatform
-            include_assessment_with_configure_patching = (patch_operation_requested == Constants.CONFIGURE_PATCHING.lower() and execution_config.assessment_mode == Constants.AssessmentModes.AUTOMATIC_BY_PLATFORM)
-            if execution_config.exec_auto_assess_only or patch_operation_requested != Constants.CONFIGURE_PATCHING.lower() or include_assessment_with_configure_patching:
-                patch_assessment_successful = patch_assessor.start_assessment()
+            # Assessment happens for all operations. If the goal seeking tracked operation is CP, then its final status can only be written after assessment reaches a terminal state
+            patch_assessment_successful = patch_assessor.start_assessment()
+            if not execution_config.exec_auto_assess_only and patch_operation_requested == Constants.CONFIGURE_PATCHING.lower():
+                configure_patching_processor.set_configure_patching_final_overall_status()
```
Comment (mirichmo) [inline src/core/src/CoreMain.py:80]: If assessment throws (raises?) an exception, does the except ensure a final status is written?

Comment (kjohn-msft) [reply src/core/src/CoreMain.py:80]: Assessment has built in full-cover exception handling with retries, and there's status reconciliation in the highest-level exception handling, but there's still a slim possibility of a super-fundamental problem (like being unable to log something basic that is never expected to fail and would need our attention anyway) that could throw this off. Evaluating if something additional should be done for that.

File: src/core/src/core_logic/ConfigurePatchingProcessor.py [hunk 3]
```diff
@@ -47,16 +48,28 @@ def start_configure_patching(self):
...
+            # If the tracked operation is Configure patching, we cannot write a final status until assessment has also written a final status (mitigation for a CRP bug)
+            if self.execution_config.operation.lower() != Constants.CONFIGURE_PATCHING.lower():
+                self.set_configure_patching_final_overall_status()
         except Exception as error:
             self.current_auto_assessment_state = Constants.AutoAssessmentStates.ERROR
-            self.__report_consolidated_configure_patch_status(status=Constants.STATUS_ERROR, error=error)
+            self.configure_patching_exception_error = error
+            # If the tracked operation is Configure patching, we cannot write a final status until assessment has also written a final status (mitigation for a CRP bug)
+            if self.execution_config.operation != Constants.CONFIGURE_PATCHING.lower():
+                self.__report_consolidated_configure_patch_status(status=Constants.STATUS_ERROR, error=self.configure_patching_exception_error)
```
Comment (rane-rajasi) [inline src/core/src/core_logic/ConfigurePatchingProcessor.py:59]: error=repr(self.configure_patching_exception_error) or convert to string?

**Comment by kjohn-msft** [comment]
@rane-rajasi  -- leaving as a note here to holistically take in this PR:
image

File: src/core/src/CoreMain.py (hunk 2 above)
Comment (kjohn-msft) [reply src/core/src/CoreMain.py]: Addressed with more safety checks.

File: src/core/src/core_logic/ConfigurePatchingProcessor.py (hunk 1 above)
Comment (kjohn-msft) [reply src/core/src/core_logic/ConfigurePatchingProcessor.py:72]: Yes, this covered, it's a coverage check failure on tooling.

**Comment by kjohn-msft** [comment]
> @rane-rajasi -- leaving as a note here to holistically take in this PR: image

This was addressed too.

**Review by feng-j678 — APPROVED** [review]

File: src/core/src/CoreMain.py [hunk 4]
```diff
@@ -74,10 +74,18 @@ def __init__(self, argv):
...
+            # Assessment happens for all operations. If the goal seeking tracked operation is CP, then its final status can only be written after assessment reaches a terminal state
+            assessment_exception_error = None
+            try:
                 patch_assessment_successful = patch_assessor.start_assessment()
+            except Exception as error:
+                assessment_exception_error = error  # hold this until configure patching is closed out
+
+            # close out configure patching if needed & then raise any 'uncontrolled' assessment exception if it occurred
+            if not execution_config.exec_auto_assess_only and patch_operation_requested == Constants.CONFIGURE_PATCHING.lower():
+                configure_patching_processor.set_configure_patching_final_overall_status()  # guarantee configure patching status write prior to throwing on any catastrophic assessment error
```
Comment (rane-rajasi) [inline src/core/src/CoreMain.py:86]: Won't CRP stop reading, if it gets this status update before it gets assessment error?

**Review by rane-rajasi — CHANGES_REQUESTED** [review]
Left one comment

File: src/core/src/CoreMain.py (hunk 4 above)
Comment (rane-rajasi) [reply src/core/src/CoreMain.py:86]: Even with retries, if all the re-attempts fail, there could still be the issue, where line 86 updates CP status, which is read by agent and hence CRP before line 88 executes, causing CRP to to think that is the terminal status, where CPSummary is Success but PatchAssessmentSummary is Transitioning

Comment (kjohn-msft) [reply src/core/src/CoreMain.py:86]: As discussed, covered in PatchAssessor.

**Review by rane-rajasi — APPROVED** [review]

**Pull request merged**